### PR TITLE
Remove mirrors

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: lib/truffle/stringio.rb [hunk 1]
```diff
@@ -221,17 +221,16 @@ def write(str)
...
       string.byte_append str
       d.pos = string.bytesize
     else
       stop = string.bytesize - pos
       if str.bytesize < stop
         stop = str.bytesize
       end
-      m = Truffle::Mirror.reflect string
-      m.splice pos, stop, str
+      Truffle.invoke_primitive(:string_splice, string, str, pos, stop, str.encoding)
```
Comment (eregon) [inline lib/truffle/stringio.rb]: Is it intentional to take the encoding of the replacement here and not `string.encoding`?
Could the primitive just get the encoding out of the source String's Rope?

File: src/main/ruby/core/string.rb [hunk 2]
```diff
@@ -1359,8 +1339,7 @@ def index(str, start=undefined)
     if str.kind_of? Regexp
       Truffle::Type.compatible_encoding self, str
 
-      m = Truffle::Mirror.reflect self
-      start = m.character_to_byte_index start
+      start = Truffle.invoke_primitive(:string_character_byte_index, self, start)
```
Comment (eregon) [inline src/main/ruby/core/string.rb]: I prefer with the "to" like `:string_character_to_byte_index` so it's clear it's conversion from what to what.
I.e., it takes an integer index as input, not a character like `string_character_index`.

File: src/main/ruby/core/range.rb [hunk 3]
```diff
@@ -243,19 +243,17 @@ def min
 
   def step_internal(step_size=1) # :yields: object
     return to_enum(:step, step_size) do
-      m = Truffle::Mirror::Range.reflect(self)
-      m.step_iterations_size(*m.validate_step_size(self.begin, self.end, step_size))
+      Truffle::RangeOperations.step_iterations_size(self, *Truffle::RangeOperations.validate_step_size(self.begin, self.end, step_size))
```
Comment (eregon) [inline src/main/ruby/core/range.rb]: That's quite a long line, maybe extract the result of validate_step_size to a local variable?
Or maybe another helper in `Truffle::RangeOperations` can be added like `validate_step_iterations_size`.

File: src/main/ruby/core/truffle/process_operations.rb [hunk 4]
```diff
@@ -0,0 +1,520 @@
+# Copyright (c) 2015, 2018 Oracle and/or its affiliates. All rights reserved. This
```
Comment (eregon) [inline src/main/ruby/core/truffle/process_operations.rb]: Looks like process_mirror.rb was not removed.

File: src/main/ruby/core/truffle/proc_operations.rb [hunk 5]
```diff
@@ -25,22 +33,22 @@
...
+          curry executable, all_args, arity
+        else
+          executable[*all_args]
         end
       end
     end
+    
   end
-end
+end
```
Comment (eregon) [inline src/main/ruby/core/truffle/proc_operations.rb]: Really minor, but I think we should always have a final newline.
Maybe there is a cop for that in RuboCop.
Anyway, doesn't for this PR.

File: src/main/ruby/core/string.rb (hunk 2 above)
Comment (eregon) [reply src/main/ruby/core/string.rb]: I see, that was the name of the primitive before as well.
I think it's a good occasion to rename it, but it doesn't have to be in this PR.

File: src/main/ruby/core/truffle/process_operations.rb (hunk 4 above)
Comment (nirvdrum) [reply src/main/ruby/core/truffle/process_operations.rb]: Good catch.

File: src/main/ruby/core/truffle/proc_operations.rb (hunk 5 above)
Comment (nirvdrum) [reply src/main/ruby/core/truffle/proc_operations.rb]: I haven't seen this as significant for quite a while. But I have no problem adding them.

File: src/main/ruby/core/string.rb (hunk 2 above)
Comment (nirvdrum) [reply src/main/ruby/core/string.rb]: I'm going to take a pass over the String primitives in another PR.

**Pull request merged**